swap/protocol_test.go

```diff
@@ -315,3 +320,150 @@ func TestTriggerDisconnectThreshold(t *testing.T) {
 		t.Fatalf("Expected still no cheque, but there are %d", lenCheques)
 	}
 }
+
+// TestSwapRPC tests some basic things over RPC
+// We want this so that we can check the API works
+func TestSwapRPC(t *testing.T) {
+
+	var (
+		p2pPort = 30100
+		ipcPath = ".swarm.ipc"
+		err     error
+	)
+
+	swap, testDir := newTestSwap(t)
+	defer os.RemoveAll(testDir)
+
+	// need to have a dummy contract or the call will fail at `GetParams` due to `NewAPI`
+	swap.contract, err = contract.InstanceAt(common.Address{}, swap.backend)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// start a service stack
+	stack := createAndStartSvcNode(swap, ipcPath, p2pPort, t)
+	defer stack.Stop()
+
+	// connect to the servicenode RPCs
+	rpcclient, err := rpc.Dial(filepath.Join(stack.DataDir(), ipcPath))
+	if err != nil {
+		t.Fatal("connect to servicenode IPC fail", "err", err)
+	}
+	defer os.RemoveAll(stack.DataDir())
+
+	// create dummy peers so that we can artificially set balances and query
+	dummyPeer1 := newDummyPeer()
+	dummyPeer2 := newDummyPeer()
+	id1 := dummyPeer1.ID()
+	id2 := dummyPeer2.ID()
+
+	// set some fake balances
+	fake1 := int64(234)
+	fake2 := int64(-100)
+
+	// query a first time, should be zero
+	var balance int64
+	err = rpcclient.Call(&balance, "swap_balance", id1)
+	if err != nil {
+		t.Fatal("servicenode RPC failed", "err", err)
+	}
+	log.Debug("servicenode balance", "balance", balance)
+
+	if balance != 0 {
+		t.Fatal("Expected balance to be 0 but it is not")
+	}
+
+	// now artificially assign some balances
+	swap.balances[id1] = fake1
```

Comment:
Should we add these values directly to `balances`, or use `Add`? (not sure myself).

Reply:
Good question. I prefer directly, as this is a test, and we want to minimize other problems, it's not a unit test for `Add`.

Reply:
To be honest, I'm undecided.

On the one hand–as you point out–we want to exclude any external factors in this test, so that if there's a problem here, it's only related to the feature at hand. This seems logical and part of good unit testing.

But on the other hand, we might have business logic in `Add` that we want to include here, and we're sort of circumventing the design by not calling it.

Since I can't decide on which one I prefer, I am OK with whatever option you choose :+1:

Reply:
Thanks. I still vote to keep it; even if you are right with the business logic on `Add`, we are actually interested in the final value for this test, not the actual business logic, thus I vote to keep to set it manually.